_posts/2018-10-02-auto-tune-all.md

```diff
@@ -0,0 +1,198 @@
+
+# Benchmark Results
+We pre-tuned some popular networks on our device cluster and released the following benchmark.
+Instructions for reproduction is at the end of this blog.
+
+Getting comprehensive benchmark results of TVM is easy since we have a unified runtime interface,
+but collecting these benchmarks on other libraries for all platforms is really a pain.
+So I put all our numbers in a table, and then do some incomplete comparison with other libraries.
+
+## Comparison
+By comparing with heavily optimized traditional libraries on each platform,
+we can make sure that the performance of our automatic compilation stack is state-of-the-art.
+
+We tested popular image classification networks on ImageNet (3x224x224) dataset with batch size = 1 and data type = float32.
+The reported numbers are time costs per image in millisecond.
+
+### ARM CPU
+
+We choose [NCNN](https://github.com/Tencent/ncnn), a widely used, hand-optimized kernel library as baseline.
+It is well-optimized by NEON assembly. For example, the engineers write
+[13k lines of code](https://github.com/Tencent/ncnn/blob/master/src/layer/arm/convolution_3x3.h) for only 3x3 convolution layers.
+Since they released benchmark number in their github repo, we directly use these numbers.
+TVM outperforms it for all networks on Rasbperry Pi 3B.
+
+![image](/images/autotune-all/arm.png){: width="90%"}
+
+### Mali GPU
+
+[ARM Compute Library](https://github.com/ARM-software/ComputeLibrary) is a vendor provide library that supports Mali GPU (OpenCL) well.
+According to the results, we are better at resnet and mobilenet since we are good at convolution layers.
+We fall behind a bit on vgg-16 because vgg-16 is an old and huge network and has several large dense layers.
+However, such kind of architecture is not suitable for mobile platforms.
+
+![image](/images/autotune-all/mali.png){: width="90%"}
+
+
+### NVIDIA GPU
+
+On NVIDIA GPU, [CuDNN](https://developer.nvidia.com/cudnn) and [TensorRT](https://developer.nvidia.com/tensorrt) are two vendor-provided libraries for training and inference respectively. Since we focus on inference,
+we run our benchmark on unbatched setting. Another tensor compiler [PlaidML](https://github.com/plaidml/plaidml) is also reported as baseline.
+We get the results from its benchmark tool [PlaidBench](https://github.com/plaidml/plaidbench).
+
+According to the results, we reaches the TensorRT level. It is really a strong baseline as we all know how NVIDIA dominates deep learning hardware.
+
+![image](/images/autotune-all/nvidia.png){: width="90%"}
+
+### AMD GPU
+
+Finally let we take a look at AMD GPU. TVM supports OpenCL and [ROCm](https://rocm.github.io/) backend. We found ROCm is better since
+it is more specialized for AMD GPUs. In terms of baseline, [MIOpen](https://github.com/ROCmSoftwarePlatform/MIOpen) is a vendor provided
+kernel library. We integrate its kernel implementation in TVM graph runtime.
```

Comment:
TVM's graph runtime integrates its kernel implementations (maybe clarify that this is optional and not relied upon for generating optimized code)